File: macos/Gutenkern/Sources/GutenkernCore/CopyBurst.swift
import Foundation

public struct CopyBurst: Equatable, Sendable {
    public static let window: TimeInterval = 1

    public private(set) var keys: [String] = []
    private var lastTime: Date?

    public init() {}

    public mutating func reset() {
        keys = []
        lastTime = nil
    }

    @discardableResult
    public mutating func add(key: String, at time: Date) -> [String] {
        if let lastTime, time.timeIntervalSince(lastTime) >= Self.window {
            keys = []
        }
        if !keys.contains(key) {
            keys.append(key)
        }
        lastTime = time
        return keys
    }
}

File: windows/Gutenkern.Core/CopyBurst.cs
namespace Gutenkern;

public sealed class CopyBurst
{
    public static readonly TimeSpan Window = TimeSpan.FromSeconds(1);

    private readonly List<string> _keys = [];
    private DateTime? _lastTime;

    public IReadOnlyList<string> Keys => _keys;

    public void Reset()
    {
        _keys.Clear();
        _lastTime = null;
    }

    public IReadOnlyList<string> Add(string key, DateTime time)
    {
        if (_lastTime is { } last && time - last >= Window)
        {
            _keys.Clear();
        }

        if (!_keys.Contains(key))
        {
            _keys.Add(key);
        }

        _lastTime = time;
        return _keys.ToArray();
    }
}
